File: app/src/main/java/com/vincent/githubusersearch/paging/UserDataSourceFactory.java
package com.vincent.githubusersearch.paging;

import androidx.annotation.NonNull;
import androidx.paging.DataSource;

import com.vincent.githubusersearch.callbacks.PagingStatusCallback;
import com.vincent.githubusersearch.model.User;

public class UserDataSourceFactory extends DataSource.Factory<Integer, User> {

    private final String keyword;
    private final PagingStatusCallback statusCallback;

    UserDataSourceFactory(String keyword, PagingStatusCallback statusCallback) {
        this.keyword = keyword;
        this.statusCallback = statusCallback;
    }

    @NonNull
    @Override
    public DataSource<Integer, User> create() {
        return new GitHubUserDataSource(keyword, statusCallback);
    }
}

File: app/src/main/java/com/vincent/githubusersearch/paging/UserDataSourceFactory.kt
package com.vincent.githubusersearch.paging

import androidx.paging.DataSource
import com.vincent.githubusersearch.callbacks.PagingStatusCallback
import com.vincent.githubusersearch.model.User

/**
 * DataSourceFactory，在這裡產生 [UserDataSource]
 */
class UserDataSourceFactory internal constructor(
        private val keyword: String,
        private val statusCallback: PagingStatusCallback?) : DataSource.Factory<Int, User>() {

    override fun create(): DataSource<Int, User> {
        return UserDataSource(keyword, statusCallback)
    }

}